### PR TITLE
[QSP-4 Fix] Double Minting Exploit

### PR DESCRIPTION
**Pull request by snowsledge**

### CODE REVIEW

File: contracts/xsgd_contract.scilla [hunk 1]
```diff
@@ -370,11 +387,10 @@ transition lawEnforcementWipingBurn(address : ByStr20, initiator : ByStr20)
       | Some b =>
         balances[address] := zero;
         currentSupply <- totalSupply;
-        new_supply = builtin sub currentSupply bal;
+        new_supply = builtin sub currentSupply b;
```
Comment (junhaoyap) [inline contracts/xsgd_contract.scilla]:
```suggestion
        new_supply = builtin sub currentSupply bal;
```

File: contracts/xsgd_contract.scilla [hunk 2]
```diff
@@ -225,33 +224,47 @@ transition configureMinter(minter : ByStr20, minterAllowedAmount : Uint128, init
...
+    match or_res with
     | False =>
-      e = { _eventname : "removeMinter FailedNotMasterMinter"; minter : minter; sender : initiator};
+      get_mint_allowance <- mintAllowed[minter];
+      mint_allowance =
+        match get_minter_allowance with
+        | Some allowance => allowance
+        | None => zero
+        end;
+      new_mint_allowance = builtin sub mint_allowance amount;
```
Comment (jjcnn) [inline contracts/xsgd_contract.scilla]: Underflow problem here.

**Review by jjcnn — CHANGES_REQUESTED** [review]

File: contracts/xsgd_contract.scilla (hunk 2 above)
Comment (junhaoyap) [inline contracts/xsgd_contract.scilla]:
```suggestion
      new_mint_allowance = builtin sub mint_allowance amount;
      new_mint_allowance =
        let amount_le_allowance = uint128_le amount mint_allowance in
        match amount_le_allowance with
        | True => builtin sub mint_allowance amount
        | False => zero
        end;
```

Comment (junhaoyap) [reply contracts/xsgd_contract.scilla]: Have noted from @AmritKumar that Scilla's integer arithmetic implement Safe Math, that would result in a revert of transaction when set below 0

At the same time @jjcnn's comment on https://github.com/Xfers/XSGD-scilla/pull/10 that it should likely end as 0 if we do 10 - 30, effectively removing the address as a minter is correct so will make the change for their behaviour to be consistent

Comment (AmritKumar) [reply contracts/xsgd_contract.scilla]: We need to delete this line. As it duplicates the logic that follows.

File: contracts/xsgd_contract.scilla [hunk 3]
```diff
@@ -225,33 +224,53 @@ transition configureMinter(minter : ByStr20, minterAllowedAmount : Uint128, init
...
-    | True =>
-      delete minters[minter];
-      minterAllowed[minter] := zero;
-      e = { _eventname : "MinterRemoved"; minter : minter; sender : initiator};
-      event e
+    or_res = let tmp = negb isMasterMinter in orb isPaused tmp;
+    match or_res with
     | False =>
-      e = { _eventname : "removeMinter FailedNotMasterMinter"; minter : minter; sender : initiator};
+      get_mint_allowance <- mintAllowed[minter];
```
Comment (AmritKumar) [inline contracts/xsgd_contract.scilla]:
```suggestion
      get_minter_allowance <- mintAllowed[minter];
```

Comment (AmritKumar) [inline contracts/xsgd_contract.scilla]: This should be `get_minter_allowance`.

File: contracts/xsgd_contract.scilla [hunk 4]
```diff
@@ -225,33 +224,53 @@ transition configureMinter(minter : ByStr20, minterAllowedAmount : Uint128, init
     or_res = let tmp = negb isMasterMinter in orb isPaused tmp;
     match or_res with
     | False =>
-      minters[minter] := one;
-      minterAllowed[minter] := minterAllowedAmount;
-      e = { _eventname : "MinterConfigured"; minter : minter; amount : minterAllowedAmount; sender : initiator};
+      get_mint_allowance <- mintAllowed[minter];
```
Comment (AmritKumar) [inline contracts/xsgd_contract.scilla]: This should be `get_minter_allowance`.

**Review by AmritKumar — APPROVED** [review]

**Pull request merged**